### PR TITLE
mNAT + LB support for swift containers

### PR DESCRIPTION
**Pull request by vakalapa**
**Reason for Change**:
mNAT + LB support changes in CNS

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [x] adds unit tests

**Notes**:

### CODE REVIEW

File: cns/restserver/internalapi.go [hunk 1]
```diff
@@ -254,3 +262,57 @@ func (service *HTTPRestService) CreateOrUpdateNetworkContainerInternal(req cns.C
...
+			nodeID, infraVnet, dncEP)
+	}
+
+	nodeRegisterRequest.NmAgentSupportedApis = supportedApis
+	if err := json.NewEncoder(&body).Encode(nodeRegisterRequest); err != nil {
+		log.Errorf("encoding json failed with %v", err)
+		return
+	}
+
+	for sleep := true; err != nil; sleep = true {
```
Comment (matmerr) [inline cns/restserver/internalapi.go:294]: I know this is just moved code :) but I think we can use ticker to avoid sleeps and returning an error after a certain amount of time
```
ticker := time.NewTicker(time.Second * 5)
for {
  select {
  case <-ticker.C:
  	response, err = httpc.Post(url, "application/json", &body)
		if err == nil {
...
...
  case <-time.After(5 * time.Minute):
  	logger.errorf("no response from nmAgent after an hour") 
        return...
  }
}
```

File: cns/restserver/api.go [hunk 2]
```diff
@@ -1415,3 +1415,45 @@ func (service *HTTPRestService) deleteHostNCApipaEndpoint(w http.ResponseWriter,
...
+	)
+
+	err = service.Listener.Decode(w, r, &req)
+	logger.Request(service.Name, &req, err)
+	if err != nil {
+		return
+	}
+
+	switch r.Method {
+	case "POST":
```
Comment (matmerr) [inline cns/restserver/api.go:1438]: could use `http.MethodPost` instead of the string

File: cns/restserver/api.go [hunk 3]
```diff
@@ -1415,3 +1415,45 @@ func (service *HTTPRestService) deleteHostNCApipaEndpoint(w http.ResponseWriter,
...
+		supportedApis, returnMessage = nmagentclient.GetNmAgentSupportedApis(
+			req.GetNmAgentSupportedApisURL)
+		if returnMessage != "" {
+			returnCode = NmAgentSupportedApisError
+			logger.Errorf("[Azure-CNS] %s", returnMessage)
+		}
+
+	default:
+		returnMessage = "[Azure-CNS] GetHostLocalIP API expects a GET."
```
Comment (matmerr) [inline cns/restserver/api.go:1447]: doesn't this API expect a `POST`?

File: cns/nmagentclient/nmagentclient.go [hunk 4]
```diff
@@ -86,3 +96,41 @@ func GetNetworkContainerVersion(
...
+	)
+
+	if getNmAgentSupportedApisURL == "" {
+		getNmAgentSupportedApisURL = fmt.Sprintf(
+			GetNmAgentSupportedApiURLFmt, WireserverIP)
+	}
+
+	response, err := common.GetHttpClient().Get(getNmAgentSupportedApisURL)
+	if err != nil || response.StatusCode != http.StatusOK || response == nil {
+		returnMessage = fmt.Sprintf("Failed to retrieve Supported Apis from NMAgent")
```
[the remark this answers is not in the document]
Comment (vakalapa) [reply cns/nmagentclient/nmagentclient.go:116]: Agreed!

File: cns/restserver/api.go (hunk 3 above)
Comment (vakalapa) [reply cns/restserver/api.go:1447]: Right, I modeled it as GET first, changed it to POST but forgot to update this message.

File: cns/nmagentclient/nmagentclient.go [hunk 5]
```diff
@@ -86,3 +96,42 @@ func GetNetworkContainerVersion(
 		networkContainerID, response, err)
 	return response, err
 }
+
+// GetNmAgentSupportedApis :- Retrieves Supported Apis from NMAgent
+func GetNmAgentSupportedApis(
+	getNmAgentSupportedApisURL string) ([]string, string) {
```
Comment (ramiro-gamarra) [inline cns/nmagentclient/nmagentclient.go:102]: This function should return an error instead of a message. I recommend reading: https://github.com/golang/go/wiki/CodeReviewComments#in-band-errors

File: cns/nmagentclient/nmagentclient.go [hunk 6]
```diff
@@ -86,3 +96,42 @@ func GetNetworkContainerVersion(
...
+	var (
+		returnMessage string
+		supportedApis []string
+	)
```
Comment (ramiro-gamarra) [inline cns/nmagentclient/nmagentclient.go:106]: In general, there's no need to declare variables far ahead of their usage in Go. `returnMessage` should really be an error that can be inlined in the returns. You should also consider returning explicit zero values where possible. That makes it easy to understand which return paths are happy path. Example:

```go
return nil, fmt.Errorf("failed to retrieve from nmagent: %v", err)
```
is a concise way to wrap an error as well as return the zero value of a slice in non-happy path. In the same manner
```go
return xmlDoc.SupportedApis, nil
```
also makes it clear that no error is being returned to the calling code.

File: cns/nmagentclient/nmagentclient.go [hunk 7]
```diff
@@ -86,3 +96,42 @@ func GetNetworkContainerVersion(
...
+	if getNmAgentSupportedApisURL == "" {
+		getNmAgentSupportedApisURL = fmt.Sprintf(
+			GetNmAgentSupportedApiURLFmt, WireserverIP)
+	}
+
+	response, err := common.GetHttpClient().Get(getNmAgentSupportedApisURL)
+	if err != nil || response.StatusCode != http.StatusOK || response == nil {
+		returnMessage = fmt.Sprintf(
+			"Failed to retrieve Supported Apis from NMAgent with error %v",
+			err.Error())
```
Comment (ramiro-gamarra) [inline cns/nmagentclient/nmagentclient.go:117]: The OR check means the error could still be nil, and this could panic because of nil dereferencing

File: cns/nmagentclient/nmagentclient.go [hunk 8]
```diff
@@ -86,3 +96,42 @@ func GetNetworkContainerVersion(
...
+		logger.Errorf("[Azure-CNS] %s", returnMessage)
```
Comment (ramiro-gamarra) [inline cns/nmagentclient/nmagentclient.go:118]: Seems like you can just let the caller code log the errors instead of logging before every return

File: cns/restserver/internalapi.go [hunk 9]
```diff
@@ -254,3 +262,57 @@ func (service *HTTPRestService) CreateOrUpdateNetworkContainerInternal(req cns.C
 	return returnCode
 
 }
+
+// Try to register node with DNC when CNS is started in managed DNC mode
+func RegisterNode(httpRestService cns.HTTPService, dncEP, infraVnet, nodeID string) {
```
Comment (ramiro-gamarra) [inline cns/restserver/internalapi.go:267]: This should return an error

File: cns/api.go [hunk 10]
```diff
@@ -212,3 +213,12 @@ type DeleteHostNCApipaEndpointRequest struct {
 type DeleteHostNCApipaEndpointResponse struct {
 	Response Response
 }
+
+type NmAgentSupportedApisRequest struct {
+	GetNmAgentSupportedApisURL string
```
Comment (ramiro-gamarra) [inline cns/api.go:218]: Curious, who will be making this request? Why do they need to pass a URL?

File: cns/restserver/api.go [hunk 11]
```diff
@@ -1415,3 +1415,44 @@ func (service *HTTPRestService) deleteHostNCApipaEndpoint(w http.ResponseWriter,
...
+	)
+
+	err = service.Listener.Decode(w, r, &req)
+	logger.Request(service.Name, &req, err)
+	if err != nil {
+		return
+	}
+
+	switch r.Method {
+	case http.MethodPost:
```
Comment (ramiro-gamarra) [inline cns/restserver/api.go:1437]: Could this be a GET? After all, the client is getting the supported APIs from CNS, and CNS knows how to communicate to nmagent. Do we need to pass a url in the body?

**Review by ramiro-gamarra — COMMENTED** [review]
Some comments/questions

File: cns/restserver/internalapi.go [hunk 12]
```diff
@@ -254,3 +262,60 @@ func (service *HTTPRestService) CreateOrUpdateNetworkContainerInternal(req cns.C
...
+
+	nodeRegisterRequest.NumCPU = numCPU
+	supportedApis, msg := nmagentclient.GetNmAgentSupportedApis(httpc, "")
+
+	if msg != "" {
+		logger.Printf("[Azure CNS] Failed to retrieve SupportedApis from NMagent of node %s with Infrastructure Network: %s PrivateEndpoint: %s",
+			nodeID, infraVnet, dncEP)
+	}
+
+	nodeRegisterRequest.NmAgentSupportedApis = supportedApis
```
Comment (aegal) [inline cns/restserver/internalapi.go:288]: Out of curiosity. Why does DNC need this information from nmagent when the node is registered?

File: cns/restserver/internalapi.go [hunk 13]
```diff
@@ -132,6 +135,11 @@ func (service *HTTPRestService) SyncNodeStatus(dncEP, infraVnet, nodeID string,
 	service.saveState()
 	service.Unlock()
 
+	if nodeInfoResponse.NmAgentApisMissing {
```
Comment (aegal) [inline cns/restserver/internalapi.go:138]: This doesn't seem like it is needed. CNS would not get past start up if register node did not succeed in the main, so this information would be there if the node was registered with DNC earlier or not?

Comment (vakalapa) [reply cns/restserver/internalapi.go:138]: This will work as kind of a re-try mechanism. If DNC fails to set its properties due to some error, in the next NodeSyncStatus it can solicit for these details again.

File: cns/restserver/internalapi.go (hunk 12 above)
Comment (vakalapa) [reply cns/restserver/internalapi.go:288]: DNC needs this information to assign GRE key to a NC's primary IP Allocation. 2 conditions are required to be satisfied for DNC to start allocating GRE keys 1. NMAgent needs to have the support, 2. No other NC without greKey allocations to be present in the Node.

File: cns/restserver/api.go (hunk 11 above)
Comment (vakalapa) [reply cns/restserver/api.go:1438]: We had to change from IMDS to Wireserver, what if we want to change this again, albeit long odds.

File: cns/api.go (hunk 10 above)
Comment (vakalapa) [reply cns/api.go:218]: DNC will be making this request.

File: cns/nmagentclient/nmagentclient.go (hunk 7 above)
Comment (vakalapa) [reply cns/nmagentclient/nmagentclient.go:118]: Agreed, good catch.

File: cns/nmagentclient/nmagentclient.go (hunk 6 above)
Comment (vakalapa) [reply cns/nmagentclient/nmagentclient.go:107]: Noted.

File: cns/NetworkContainerContract.go [hunk 14]
```diff
@@ -343,4 +343,11 @@ func (networkContainerRequestPolicy *NetworkContainerRequestPolicies) Validate()
 type NodeInfoResponse struct {
 	NetworkContainers  []CreateNetworkContainerRequest
 	GetNCVersionURLFmt string
+	NmAgentApisMissing bool
```
[the remark this answers is not in the document]
Comment (vakalapa) [reply cns/NetworkContainerContract.go:346]: Right now i do not think there are any other requests or actions apart from the one in question. SO kept it simple with a single flag.

File: cns/restserver/internalapi.go (hunk 13 above)
Comment (aegal) [reply cns/restserver/internalapi.go:138]: The only thing I don't quite understand is if DNC fails to set this properties, then the register node call should fail, so we would never reach here.

Comment (vakalapa) [reply cns/restserver/internalapi.go:138]: Today, register Node happens only once, at the time of initial CNS bringup, main.go L#385. And yes, we do not want to register if DNC sets this flag as false (same as todays behavior)

File: cns/NetworkContainerContract.go [hunk 15]
```diff
@@ -343,4 +343,11 @@ func (networkContainerRequestPolicy *NetworkContainerRequestPolicies) Validate()
...
+}
+
+// NodeRegisterRequest - Struct to hold the node register request.
+type NodeRegisterRequest struct {
+	NumCPU               int
+	NmAgentSupportedApis []string
```
[the remark this answers is not in the document]
Comment (vakalapa) [reply cns/NetworkContainerContract.go:352]: ideally first NodeRegister should look like 
NodeRegisterRequest  { NumCPU : 2 , NmAgentSupportedApis : [ "Api1", "Api2"] }

any subsequent Register will be 
NodeRegisterRequest  { NumCPU : 2 , NmAgentSupportedApis : [ ""] }

File: cns/restserver/internalapi.go (hunk 13 above)
Comment (ramiro-gamarra) [reply cns/restserver/internalapi.go:138]: I don't follow either. If I'm reading main correctly, if node registration failed during start up, that terminates CNS. Registration must succeed before the sync loop is even started. Once it's registered, why do we need to try to register it again here?

File: cns/restserver/internalapi.go [hunk 16]
```diff
@@ -254,3 +265,58 @@ func (service *HTTPRestService) CreateOrUpdateNetworkContainerInternal(req cns.C
...
+		body                bytes.Buffer
+		nodeRegisterRequest cns.NodeRegisterRequest
+	)
+
+	nodeRegisterRequest.NumCPU = numCPU
+	supportedApis, retErr := nmagentclient.GetNmAgentSupportedApis(httpc, "")
+
+	if retErr != nil {
+		logger.Printf("[Azure CNS] Failed to retrieve SupportedApis from NMagent of node %s with Infrastructure Network: %s PrivateEndpoint: %s",
+			nodeID, infraVnet, dncEP)
```
Comment (ramiro-gamarra) [inline cns/restserver/internalapi.go:287]: Bubble error up. I don't think we should attempt to continue if this step failed.

File: cns/restserver/internalapi.go (hunk 1 above)
Comment (ramiro-gamarra) [reply cns/restserver/internalapi.go:296]: Agree with Mat

**Review by ramiro-gamarra — COMMENTED** [review]
Few more comments/questions

File: cns/NetworkContainerContract.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (vakalapa) [reply cns/NetworkContainerContract.go:346]: CNS will ask for regular syncNodeStatus, as a response, DNC sends back list of NCs, version and the flag for NMAgentApis list. And DNC will set this flag only if the NodeConfig has GreCapableNotSet property. If the Node is not capable or if the Node has been previously updated with this list then DNC will skip asking for this information.

File: cns/NetworkContainerContract.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (vakalapa) [reply cns/NetworkContainerContract.go:382]: Corrected this behavior, removed the re-register logic. So the requests will be idempotent.

File: cns/restserver/internalapi.go (hunk 13 above)
Comment (vakalapa) [reply cns/restserver/internalapi.go]: After thinking through and testing, i removed the re-register logic. Now if the Node needs to get the ability to add GreKeys, Cx will need to drain NCs, unregister the Node and register again.

File: cns/restserver/api.go (hunk 11 above)
Comment (vakalapa) [reply cns/restserver/api.go:1462]: DNC -> CNS POST message with payload on what CNS needs to send to wireserver. 
CNS-> Wireserver is GET

File: cns/service/main.go [hunk 17]
```diff
@@ -246,43 +248,90 @@ func printVersion() {
...
+			logger.Errorf(err.Error())
+			responseChan <- err
 		}
 
-		if sleep {
-			time.Sleep(acn.FiveSeconds)
-		}
+		response.Body.Close()
+	} else {
+		logger.Errorf("[Azure CNS] Failed to register node with err: %+v", err)
```
Comment (ramiro-gamarra) [inline cns/service/main.go:333]: Is there a reason why this path doesn't send the error? Also, i know this is just adding support of channels, but probably a good opportunity to clean up the function since its kinda hard to follow the code flow. Suggest returning early like:
```go
res, err := post()
if err != nil { /* return */ }

defer res.body.close()

if res.StatusCode != { /* return */ }

if err := decode(req); err != nil { /* return */ }

setNodeOrch(req)
```

**Review by ramiro-gamarra — COMMENTED** [review]
Few comments

File: cns/service/main.go [hunk 18]
```diff
@@ -246,43 +248,90 @@ func printVersion() {
...
+	supportedApis, retErr := nmagentclient.GetNmAgentSupportedApis(httpc, "")
+
+	if retErr != nil {
+		logger.Errorf("[Azure CNS] Failed to retrieve SupportedApis from NMagent of node %s with Infrastructure Network: %s PrivateEndpoint: %s",
+			nodeID, infraVnet, dncEP)
+		return retErr
+	}
+
+	//To avoid any null-pointer deferencing errors.
+	if supportedApis == nil {
```
[the remark this answers is not in the document]
Comment (vakalapa) [reply cns/service/main.go:271]: Okay, I wanted to insure DNC from any null pointer dereferencing.

File: cns/service/main.go [hunk 19]
```diff
@@ -246,43 +248,90 @@ func printVersion() {
...
+	responseChan := make(chan error)
+
+	for {
+		select {
+		case responseErr := <-responseChan:
+			return responseErr
+		case <-nodeRegisterTicker.C:
+			go sendRegisterNodeRequest(httpc, httpRestService, nodeRegisterRequest, url, responseChan)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (vakalapa) [reply cns/service/main.go:287]: Will correct the first 5 sec delay. Existing behavior is to retry forever. You are right, we need a better mechanism here. I can add a work item to improve this logic for exponential back off as a separate effort ?

File: cns/service/main.go (hunk 17 above)
Comment (vakalapa) [reply cns/service/main.go:333]: Okay will refactor. Yes, I am assuming this error is retriable error, like timeout or some other transport path error. (which most probably can be due to delays in setting up of azure services). The other errors are specific from DNC side, which i am attributing as not retriable.

**Review by ramiro-gamarra — APPROVED** [review]
lgtm!

File: .pipelines/e2e-step-template.yaml [hunk 20]
```diff
@@ -90,6 +90,7 @@ steps:
       export REGIONS=$(AKS_ENGINE_REGION) 
       export IS_JENKINS=false 
       export DEBUG_CRASHING_PODS=true
+      export
```
Comment (csfmomo) [inline .pipelines/e2e-step-template.yaml:93]: Why we need this one?

File: cns/api.go (hunk 10 above)
Comment (csfmomo) [reply cns/api.go:219]: Curious if DNC will making this request, why the code is in CNS side? Does that mean CNS will decode the response?

File: .pipelines/e2e-step-template.yaml (hunk 20 above)
Comment (vakalapa) [reply .pipelines/e2e-step-template.yaml:93]: To print out all the env variables. For better debuggability. This will mask any secrets like below:

```
declare -x CLIENT_ID="55990953-d723-433b-a204-01af59561ed8" 
declare -x CLIENT_SECRET="***" 
declare -x CLUSTER_DEFINITION="./cniLinux1604.json"
```
